framework/aster-frame/src/mm/space.rs

```diff
@@ -13,28 +16,41 @@ use crate::{
     arch::mm::{
         tlb_flush_addr_range, tlb_flush_all_excluding_global, PageTableEntry, PagingConsts,
     },
+    cpu::CpuExceptionInfo,
     mm::{
         page_table::{Cursor, PageTableQueryResult as PtQr},
         Frame, MAX_USERSPACE_VADDR,
     },
     prelude::*,
+    task::{current_task, disable_preempt, DisablePreemptGuard},
     Error,
 };
 
+#[allow(clippy::type_complexity)]
 /// Virtual memory space.
 ///
 /// A virtual memory space (`VmSpace`) can be created and assigned to a user space so that
 /// the virtual memory of the user space can be manipulated safely. For example,
 /// given an arbitrary user-space pointer, one can read and write the memory
-/// location refered to by the user-space pointer without the risk of breaking the
+/// location referred to by the user-space pointer without the risk of breaking the
 /// memory safety of the kernel space.
 ///
 /// A newly-created `VmSpace` is not backed by any physical memory pages.
 /// To provide memory pages for a `VmSpace`, one can allocate and map
 /// physical memory (`Frame`s) to the `VmSpace`.
-#[derive(Debug)]
+///
+/// A `VmSpace` can also attach a page fault handler, which will be invoked to handle
+/// page faults generated in user space.
 pub struct VmSpace {
     pt: PageTable<UserMode>,
+    page_fault_handler: Once<
+        Arc<
```

Comment:
To have a `dyn Fn` here we have an additional heap allocation here. Since a plain function pointer `fn(&VmSpace, &CpuExceptionInfo) -> Result<(), ()>` works well for 99% cases (at least for all usages in this PR), we might just do it to avoid this heap allocation.

Meanwhile, a `PageTable` doesn't require any heap space now (just need page allocation), avoiding heap allocations in `VmSpace` seems fantastic to me.